test/cri-containerd/clone_test.go

```diff
@@ -0,0 +1,404 @@
+	containerRequest.PodSandboxId = podID
+	containerRequest.SandboxConfig = sandboxRequest.Config
+	containerID = createContainer(t, client, ctx, containerRequest)
+	startContainer(t, client, ctx, containerID)
+	return podID, containerID
+}
+
+// Creates a template sandbox and then a template container inside it.
+// Since, template container can take time to finish the init process and then exit (at which
+// point it will actually be saved as a template) this function wait until the template is
+// actually saved.
+// It is the callers responsibility to clean the stop and remove the cloned
+// containers and pods.
+func createTemplateContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, templateSandboxRequest *runtime.RunPodSandboxRequest, templateContainerRequest *runtime.CreateContainerRequest) (templatePodID, templateContainerID string) {
+	templatePodID, templateContainerID = createPodAndContainer(ctx, t, client, templateSandboxRequest, templateContainerRequest)
+
+	// Send a context with deadline for waitForTemplateSave function
+	d := time.Now().Add(10 * time.Second)
+	ctx, cancel := context.WithDeadline(ctx, d)
+	defer cancel()
+	waitForTemplateSave(ctx, t, templatePodID)
+	return
+}
+
+// Creates a clone from the given template pod and container.
+// It is the callers responsibility to clean the stop and remove the cloned
+// containers and pods.
+func createClonedContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, templatePodID, templateContainerID string, cloneNumber int) (clonedPodID, clonedContainerID string) {
+	cloneSandboxRequest := getClonedPodConfig(cloneNumber, templatePodID)
+	cloneContainerRequest := getClonedContainerConfig(cloneNumber, templateContainerID)
+	clonedPodID, clonedContainerID = createPodAndContainer(ctx, t, client, cloneSandboxRequest, cloneContainerRequest)
+	return
+}
+
+// Runs a command inside given container and verifies if the command executes successfully.
+func verifyContainerExec(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, containerID string) {
+	execCommand := []string{
+		"ping",
+		"www.microsoft.com",
+	}
+
+	execRequest := &runtime.ExecSyncRequest{
+		ContainerId: containerID,
+		Cmd:         execCommand,
+		Timeout:     20,
+	}
+
+	r := execSync(t, client, ctx, execRequest)
+	output := strings.TrimSpace(string(r.Stdout))
+	errorMsg := string(r.Stderr)
+	exitCode := int(r.ExitCode)
+
+	if exitCode != 0 || len(errorMsg) != 0 {
+		t.Fatalf("Failed execution inside container %s with error: %s, exitCode: %d", containerID, errorMsg, exitCode)
+	} else {
+		t.Logf("Exec(container: %s) stdout: %s, stderr: %s, exitCode: %d\n", containerID, output, errorMsg, exitCode)
+	}
+}
+
+// A simple test to just create a template container and then create one
+// cloned container from that template.
+func Test_CloneContainer_WCOW(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Make sure to add the WCOW hypervisor feature to all of these tests

Reply:
Good catch. Done.